### PR TITLE
Refactor runtime exception instrumentation

### PR DESCRIPTION
**Pull request by cristina-david**
This adds some of the runtime exception instrumentation after Java bytecode conversion , just before typechecking. Also, the corresponding assertions that used to be in place before the runtime exceptions instrumentation is removed from the conversion and added during this subsequent instrumentation phase.

### CODE REVIEW

File: regression/cbmc-java/ArrayIndexOutOfBoundsException1/ArrayIndexOutOfBoundsExceptionTest.java [hunk 1]
```diff
@@ -0,0 +1,12 @@
+public class ArrayIndexOutOfBoundsExceptionTest {
```
Comment (smowton) [inline regression/cbmc-java/ArrayIndexOutOfBoundsException1/ArrayIndexOutOfBoundsExceptionTest.java]: Dodgy indentation, use spaces everywhere

File: regression/cbmc-java/ClassCastException1/ClassCastExceptionTest.java [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+
+class B {
+  int j;
+}
+
+public class ClassCastExceptionTest {
+  public static void main(String args[]) {
+      try {
+          Object x = new Integer(0);
+          String y=(String)x;
```
Comment (smowton) [inline regression/cbmc-java/ClassCastException1/ClassCastExceptionTest.java]: style

File: regression/cbmc-java/ClassCastException2/ClassCastExceptionTest.java [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+
+class B {
+  int j;
+}
+
+public class ClassCastExceptionTest {
+  public static void main(String args[]) {
+      try {
+          Object x = new Integer(0);
+          String y=(String)x;
```
Comment (smowton) [inline regression/cbmc-java/ClassCastException2/ClassCastExceptionTest.java]: style

File: regression/cbmc-java/ClassCastException2/test.desc [hunk 4]
```diff
@@ -0,0 +1,9 @@
+CORE
```
Comment (smowton) [inline regression/cbmc-java/ClassCastException2/test.desc]: Aren't tests CCE1 and CCE2 identical?

File: regression/cbmc-java/NullPointerException2/Test.java [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+
+class A {
+    int i;
+}
+
+public class Test {
+    public static void main(String args[]) {
+        A a=null;
+        try {
+	    a.i=0;
```
Comment (smowton) [inline regression/cbmc-java/NullPointerException2/Test.java]: tabs

File: regression/cbmc-java/NullPointerException2/Test.java [hunk 6]
```diff
@@ -0,0 +1,18 @@
...
+            throw new B();
```
Comment (smowton) [inline regression/cbmc-java/NullPointerException2/Test.java]: Is there some importance in these unreachable throws? If so add a comment explaining what they're for

File: regression/cbmc-java/NullPointerException3/Test.java [hunk 7]
```diff
@@ -0,0 +1,18 @@
+class B extends RuntimeException {}
+
+class A {
```
Comment (smowton) [inline regression/cbmc-java/NullPointerException3/Test.java]: How does this test differ from NPE2?

File: src/cbmc/cbmc_parse_options.h [hunk 8]
```diff
@@ -64,7 +64,8 @@ class optionst;
   "(round-to-nearest)(round-to-plus-inf)(round-to-minus-inf)(round-to-zero)" \
   "(graphml-witness):" \
   "(java-max-vla-length):(java-unwind-enum-static)" \
-  "(java-cp-include-files):" \
+  "(java-throw-runtime-exceptions)" \
+  "(java-cp-include-files):"                   \
```
Comment (smowton) [inline src/cbmc/cbmc_parse_options.h]: No need to change this line

File: src/java_bytecode/java_bytecode_instrument.cpp [hunk 9]
```diff
@@ -0,0 +1,520 @@
...
+
+    block.copy_to_operands(code);
+    code=block;
+  }
+}
+
+/// Computes the instrumentation for `expr` in the form of
+/// either assertions or runtime exceptions.
+/// \par parameters: `expr`: the expression for which we compute
+/// strumentation \return: The instrumentation required for `expr`
```
Comment (smowton) [inline src/java_bytecode/java_bytecode_instrument.cpp]: typo, and believe \return needs a newline

File: src/java_bytecode/java_bytecode_instrument.cpp [hunk 10]
```diff
@@ -0,0 +1,520 @@
...
+
+  void instrument_code(exprt &expr);
+  codet instrument_expr(const exprt &expr);
+};
+
+
+/// Creates a class stub for exc_name and generates a
+///  conditional GOTO such that exc_name is thrown when
+///  cond is met.
+/// \par parameters: `cond`: condition to be meat in order
```
Comment (smowton) [inline src/java_bytecode/java_bytecode_instrument.cpp]: meat->met

File: src/java_bytecode/java_bytecode_instrument.cpp [hunk 11]
```diff
@@ -0,0 +1,520 @@
...
+/// \return Returns the code initialising the throwing the
+/// exception
+codet java_bytecode_instrumentt::throw_exception(
+  const exprt &cond,
+  const source_locationt &original_loc,
+  const irep_idt &exc_name)
+{
+  exprt exc;
+  code_blockt init_code;
+  const irep_idt &exc_obj_name=id2string(goto_functionst::entry_point())+
```
Comment (smowton) [inline src/java_bytecode/java_bytecode_instrument.cpp]: style (wrap after `=`)

File: src/java_bytecode/java_bytecode_instrument.cpp [hunk 12]
```diff
@@ -0,0 +1,520 @@
...
+    "::"+id2string(exc_name);
+
+  if(!symbol_table.has_symbol(exc_obj_name))
+  {
+    // for now, create a class stub
```
Comment (smowton) [inline src/java_bytecode/java_bytecode_instrument.cpp]: Use `java_bytecode_convert_classt::generate_class_stub`?

File: src/java_bytecode/java_bytecode_instrument.cpp [hunk 13]
```diff
@@ -0,0 +1,520 @@
...
+/// ClassCastException/generates an assertion when necessary;
+/// Exceptions are thrown when the `throw_runtime_exceptions` flag is set.
+/// Otherwise, assertions are emitted.
+/// \par parameters: `class1`: the subclass
+/// `class2`: the super class
+/// `original_loc: source location in the original code
+/// \return Based on the value of the flag `throw_runtime_exceptions`,
+/// it returns code that either throws an ClassCastException or emits an
+/// assertion checking the subtype relation
+codet java_bytecode_instrumentt::check_class_cast(
```
Comment (smowton) [inline src/java_bytecode/java_bytecode_instrument.cpp]: Could we do with a testcase that casts to a derived type? For example, if C <: B <: A, something that casts an actual C from an A to a B (should pass but doesn't precisely match the target type)

File: src/java_bytecode/java_bytecode_instrument.cpp [hunk 14]
```diff
@@ -0,0 +1,520 @@
...
+  codet &code=to_code(expr);
+
+  const irep_idt &statement=code.get_statement();
+
+  if(statement==ID_assign)
+  {
+    code_assignt &code_assign=to_code_assign(code);
+    INVARIANT(
+      code_assign.has_operands(),
+      "Assignments should have (2) operands");
```
Comment (smowton) [inline src/java_bytecode/java_bytecode_instrument.cpp]: Might as well check there are actually 2

File: src/java_bytecode/java_bytecode_instrument.cpp [hunk 15]
```diff
@@ -0,0 +1,520 @@
...
+    .set_property_class("array-create-negative-size");
+  return check;
+}
+
+/// Augments `expr` with instrumentation in the form of either
+/// assertions or runtime exceptions
+/// \par parameters: `expr`: the expression to be instrumented
+void java_bytecode_instrumentt::instrument_code(exprt &expr)
+{
+  PRECONDITION(expr.id()==ID_code);
```
Comment (smowton) [inline src/java_bytecode/java_bytecode_instrument.cpp]: Redundant with the following line

File: src/java_bytecode/java_bytecode_instrument.cpp [hunk 16]
```diff
@@ -0,0 +1,520 @@
...
+}
+
+/// Computes the instrumentation for `expr` in the form of
+/// either assertions or runtime exceptions.
+/// \par parameters: `expr`: the expression for which we compute
+/// strumentation \return: The instrumentation required for `expr`
+codet java_bytecode_instrumentt::instrument_expr(
+  const exprt &expr)
+{
+  if(expr.id()==ID_plus)
```
Comment (smowton) [inline src/java_bytecode/java_bytecode_instrument.cpp]: This pattern matching seems likely to accidentally match other expressions in the future... can we identify array accesses more concretely? How about setting a boolean flag on the ID_plus expression during convert_method (ID_java_array_access?) and recognising it here?

File: src/java_bytecode/java_bytecode_instrument.cpp [hunk 17]
```diff
@@ -0,0 +1,520 @@
...
+  {
+    // this correspond to ?astore so we check that 0<=index<array.length
```
Comment (smowton) [inline src/java_bytecode/java_bytecode_instrument.cpp]: Also ?aload

File: src/java_bytecode/java_bytecode_instrument.cpp [hunk 18]
```diff
@@ -0,0 +1,520 @@
...
+        expr.source_location(),
+        true);
+    else if(statement==ID_java_new_array)
+      // this correpond to new array so we check that
+      // length is >=0
+      return check_array_length(
+        expr.op0(),
+        expr.source_location());
+  }
+  if(expr.id()==ID_member)
```
Comment (smowton) [inline src/java_bytecode/java_bytecode_instrument.cpp]: else if

File: src/java_bytecode/java_bytecode_instrument.cpp [hunk 19]
```diff
@@ -0,0 +1,520 @@
...
+  {
+    // this corresponds to either getfield or putfield
+    // so we must check null pointer dereference
```
Comment (smowton) [inline src/java_bytecode/java_bytecode_instrument.cpp]: Note this also triggers (correctly) for ?astore, ?aload which synthesise a member_exprt, but will trigger twice (once for ->length, one for ->data). Might be simplest to have the frontend tag where null checks ought to be, in the style of the above suggested is-array-access tag?

File: regression/cbmc-java/ClassCastException2/test.desc (hunk 4 above)
Comment (cristina-david) [inline regression/cbmc-java/ClassCastException2/test.desc]: Yes, they somehow were, thanks for noticing that. I've now replaced CCE2 by the one you suggest below (with C<:B<:A), if I understood it correctly.

File: regression/cbmc-java/NullPointerException3/Test.java (hunk 7 above)
Comment (cristina-david) [inline regression/cbmc-java/NullPointerException3/Test.java]: NPE2 uses putfield, whereas NPE3 uses getfield. They ultimately rely on the same code but I still wanted to check they are both good.

File: src/java_bytecode/java_bytecode_instrument.cpp (hunk 12 above)
Comment (cristina-david) [inline src/java_bytecode/java_bytecode_instrument.cpp]: Yes, I wanted to use it but it's protected.

File: src/java_bytecode/java_bytecode_instrument.cpp (hunk 13 above)
Comment (cristina-david) [inline src/java_bytecode/java_bytecode_instrument.cpp]: This is now CCE2.

**Comment by cristina-david** [comment]
@smowton Thanks for the comments, I've addressed them.

File: src/java_bytecode/java_bytecode_instrument.cpp (hunk 12 above)
Comment (smowton) [inline src/java_bytecode/java_bytecode_instrument.cpp]: I say make it public and use it.

File: src/java_bytecode/java_bytecode_instrument.cpp [hunk 20]
```diff
@@ -0,0 +1,523 @@
...
+    code_blockt block;
+    block.copy_to_operands(instrument_expr(code.op0()));
+    block.copy_to_operands(instrument_expr(code.op1()));
+    block.copy_to_operands(code);
+
+    code=block;
+  }
+  else if(statement==ID_expression)
+  {
+    INVARIANT(
```
Comment (mgudemann) [inline src/java_bytecode/java_bytecode_instrument.cpp]: use `to_code_expression` here analogously to assignments above

File: src/java_bytecode/java_bytecode_instrument.cpp [hunk 21]
```diff
@@ -0,0 +1,523 @@
...
+  }
+  else if(statement==ID_assert)
+  {
+    INVARIANT(
+      code.has_operands(),
+      "Asserts should have (1) operands");
+    // does this correspond to checkcast?
+    if(code.op0().id()==ID_java_instanceof)
+    {
+      code_blockt block;
```
Comment (mgudemann) [inline src/java_bytecode/java_bytecode_instrument.cpp]: here should be an INVARIANT on `op0` having two operands

File: src/java_bytecode/java_bytecode_instrument.cpp [hunk 22]
```diff
@@ -0,0 +1,523 @@
...
+    block.copy_to_operands(code);
+    code=block;
+  }
+  else if(statement==ID_switch)
+  {
+    code_blockt block;
+    code_switcht &code_switch=to_code_switch(code);
+    block.copy_to_operands(
+      instrument_expr(code_switch.value()));
+    block.copy_to_operands(code);
```
Comment (mgudemann) [inline src/java_bytecode/java_bytecode_instrument.cpp]: why not `code_switch.body()`, too?

File: src/java_bytecode/java_bytecode_instrument.cpp [hunk 23]
```diff
@@ -0,0 +1,523 @@
...
+#include "java_root_class.h"
+#include "java_types.h"
+
+class java_bytecode_instrumentt
+{
+public:
+  java_bytecode_instrumentt(
+    symbol_tablet &_symbol_table,
+    bool _throw_runtime_exceptions,
+    size_t _max_array_length):
```
Comment (mgudemann) [inline src/java_bytecode/java_bytecode_instrument.cpp]: the above two could be `const`

File: src/java_bytecode/java_bytecode_instrument.cpp (hunk 12 above)
Comment (cristina-david) [inline src/java_bytecode/java_bytecode_instrument.cpp]: Done. Could you please take a look again at this PR?

**Comment by smowton** [comment]
Sorry I should've been clearer-- this way we end up having to pass `java_bytecode_instrumentt` a bunch of stuff it doesn't need, lazy-methods-mode and so on. Rather I meant let's kick `generate_class_stub` out into a public utility method and use it from both `java_bytecode_convert_classt` and `java_bytecode_instrumentt`

**Comment by cristina-david** [comment]
@smowton and @mgudemann I've done the requested changes. Can you please take another look at it?

File: src/java_bytecode/java_utils.cpp [hunk 24]
```diff
@@ -52,3 +53,38 @@ unsigned java_method_parameter_slots(const code_typet &t)
...
+  new_symbol.type=class_type;
+  new_symbol.mode=ID_java;
+  new_symbol.is_type=true;
+
+  symbolt *class_symbol;
+
+  if(symbol_table.move(new_symbol, class_symbol))
+  {
+    // warning() << "stub class symbol " << new_symbol.name
+    //           << " already exists" << eom;
```
Comment (mgudemann) [inline src/java_bytecode/java_utils.cpp]: should either be gone or activated

File: src/java_bytecode/Makefile [hunk 25]
```diff
@@ -16,6 +16,7 @@ SRC = bytecode_info.cpp \
       java_class_loader.cpp \
       java_class_loader_limit.cpp \
       java_entry_point.cpp \
+      java_bytecode_instrument.cpp \
```
Comment (mgudemann) [inline src/java_bytecode/Makefile]: please sort alphabetically

**Pull request merged**